### PR TITLE
consistent Delete/Copy/Save UX across media types

### PR DESCRIPTION
**Pull request by michaelkirk**
Delete/Copy/Save media items via long press menu from within conversation view.

Fix for #1103 and #704

The *new* functionality is 
 * delete: video, audio, photo, and animated GIF messages
 * copy: video (as mp4), photo (as jpg), and animated GIF messages (was already possible to copy plain text message)
 * save: video (as mp4), photo (as jpg), and animated GIF messages (intentionally not possible to save plain text message or audio messages)

### Motivation

This is in line with how people expect to interact with messages (e.g. see WhatsApp and the Native Messaging app), and we were already doing it with Text messages. But for some reason we were doing these actions differently for images, and not at all for videos.

### Previous UX

In order to delete an image, first, you'd tap on the image to bring up the full screen view:

image

Then, tap "Share" button to open an action sheet, then press delete:
image

### Proposed UX

Now we're doing this consistently across message types using the edit menu, via long-press:

image

Consequently the "share" icon has been removed from the fullscreen image view:
image

### CODE REVIEW

**Comment by rocknjohn** [comment]
Looks nice Michael. The native action sheets look much nicer. The DJW animations had felt unsettling too.

But wouldn't it make sense to have "Save/Copy/Delete" all available when long pressing on an image (https://github.com/WhisperSystems/Signal-iOS/issues/825), instead of only "Delete". These are all common actions.

**Comment by michaelkirk** [comment]
@rocknjohn - I agree with you. Plus that's the behavior of native messaging app. 

Here's what would need to be done:

- [ ] Move "save" action from "share" menu to "edit" menu
- [ ] Kill the "share" menu all together until there is an actual "share" action you can do with it.
- [ ] Add a "copy" action for photos edit menu
- [ ] Add a "copy" and a "save" action for videos

I won't get to it until later this week, per other working obligations. Thanks for the feedback!

**Comment by michaelkirk** [comment]
OK - some progress here. This PR is getting pretty large...

I've killed off the "share" menu all together, since the items in it are more conventionally found in the edit-menu (that's *this* guy, which you get to by long pressing).

image

The bad news is that until #1023 is complete, we won't be able to paste anything into Signal. So, silly enough, you can copy these things, but you can still only paste text into Signal at this point.

The feature feels incomplete without *also* fixing the paste situation, but this PR is already getting quite big. I'll update the description to reflect the current state of things.

File: Signal/src/view controllers/TSMessageAdapter.m [hunk 1]
```diff
@@ -237,6 +241,79 @@ - (NSDate *)date {
...
+        //TODO implement save for video
+    }
+    return NO;
+}
+
+- (void)performEditingAction:(SEL)action {
+    if (action == @selector(delete:)) {
+        DDLogDebug(@"Deleting interaction with uniqueId: %@", self.interaction.uniqueId);
+        [self.interaction remove];
+    } else if (action == @selector(copy:)) {
```
Comment (michaelkirk) [inline Signal/src/view controllers/TSMessageAdapter.m]: This is screaming for refactoring based on inheritance, but I was trying not to move too much at once.

**Comment by pstasiak** [comment]
> But wouldn't it make sense to have "Save/Copy/Delete" all available when long pressing on an image (#825), instead of only "Delete". These are all common actions.

Hey, I've just created #1127 with an update of JSQMessagesViewController with copy media functionality.

**Comment by greycubesgav** [comment]
+1 for this to be merged and committed to the App Store available verion.

**Comment by JohanDoe** [comment]
I am using Signal 2.4.0 from the app store. The share button has indeed disappeared. But I only see copy/delete using the edit menu via long-press on an image. 

Is this intentional?

**Comment by michaelkirk** [comment]
@JohanDoe - That's not intentional, and sounds like a bug.

You should be able to save anything except plain text and audio messages. Can you file a new bug and include screenshots?

**Comment by JohanDoe** [comment]
@michaelkirk strange. Created a new conversation thread for the bug report screenshots and now everything is as it used to be. Can save images, also in old thread. 

Can't reproduce right now. Will keep an eye on it and if I can reproduce I will submit a bug report.

**Comment by michaelkirk** [comment]
👀 

Thanks @JohanDoe!

**Pull request merged**